zypp: FTBFS when using '-Wall -Werror=format-security' in CFLAGS & CXXFLAGS

I attempted to build PackageKit with the zypp backend on Fedora 23 with libzypp 15.21.1, and it failed to compile.

The error output below is what happened when compiling the zypp backend code:
```
[  317s] Making all in zypp
[  317s] make[3]: Entering directory '/home/abuild/rpmbuild/BUILD/PackageKit-1.0.11/backends/zypp'
[  317s] /bin/sh ../../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../..  -I../../src -I../../lib -DPK_COMPILATION -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include  -pthread -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include  -pthread -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include   -Wno-deprecated   --std=c++0x -Wall -Woverloaded-virtual -Wnon-virtual-dtor -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -c -o libpk_backend_zypp_la-pk-backend-zypp.lo `test -f 'pk-backend-zypp.cpp' || echo './'`pk-backend-zypp.cpp
[  317s] libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../.. -I../../src -I../../lib -DPK_COMPILATION -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -pthread -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -pthread -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -Wno-deprecated --std=c++0x -Wall -Woverloaded-virtual -Wnon-virtual-dtor -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -c pk-backend-zypp.cpp  -fPIC -DPIC -o .libs/libpk_backend_zypp_la-pk-backend-zypp.o
[  322s] pk-backend-zypp.cpp: In member function 'virtual zypp::target::rpm::InstallResolvableReport::Action ZyppBackend::InstallResolvableReportReceiver::problem(zypp::Resolvable::constPtr, zypp::target::rpm::InstallResolvableReport::Error, const string&, zypp::target::rpm::InstallResolvableReport::RpmLevel)':
[  322s] pk-backend-zypp.cpp:279:97: error: format not a string literal and no format arguments [-Werror=format-security]
[  322s]    pk_backend_job_error_code (_job, PK_ERROR_ENUM_PACKAGE_FAILED_TO_INSTALL, description.c_str ());
[  322s]                                                                                                  ^
[  322s] pk-backend-zypp.cpp: In member function 'virtual zypp::target::rpm::RemoveResolvableReport::Action ZyppBackend::RemoveResolvableReportReceiver::problem(zypp::Resolvable::constPtr, zypp::target::rpm::RemoveResolvableReport::Error, const string&)':
[  322s] pk-backend-zypp.cpp:314:114: error: format not a string literal and no format arguments [-Werror=format-security]
[  322s]                  pk_backend_job_error_code (_job, PK_ERROR_ENUM_CANNOT_REMOVE_SYSTEM_PACKAGE, description.c_str ());
[  322s]                                                                                                                   ^
[  322s] pk-backend-zypp.cpp: In member function 'virtual zypp::media::MediaChangeReport::Action ZyppBackend::MediaChangeReportReceiver::requestMedia(zypp::Url&, unsigned int, const string&, zypp::media::MediaChangeReport::Error, const string&, const std::vector<std::__cxx11::basic_string<char> >&, unsigned int&)':
[  322s] pk-backend-zypp.cpp:413:90: error: format not a string literal and no format arguments [-Werror=format-security]
[  322s]    pk_backend_job_error_code (_job, PK_ERROR_ENUM_REPO_NOT_AVAILABLE, description.c_str ());
[  322s]                                                                                           ^
[  322s] pk-backend-zypp.cpp: In member function 'zypp::ZYpp::Ptr ZyppJob::get_zypp()':
[  322s] pk-backend-zypp.cpp:603:111: error: format not a string literal and no format arguments [-Werror=format-security]
[  322s]    pk_backend_job_error_code (priv->currentJob, PK_ERROR_ENUM_FAILED_INITIALIZATION, ex.asUserString().c_str() );
[  322s]                                                                                                                ^
[  322s] pk-backend-zypp.cpp:606:104: error: format not a string literal and no format arguments [-Werror=format-security]
[  322s]    pk_backend_job_error_code (priv->currentJob, PK_ERROR_ENUM_INTERNAL_ERROR, ex.asUserString().c_str() );
[  322s]                                                                                                         ^
[  322s] pk-backend-zypp.cpp: In function 'zypp::RepoInfo zypp_get_Repository(PkBackendJob*, const gchar*)':
[  322s] pk-backend-zypp.cpp:972:91: error: format not a string literal and no format arguments [-Werror=format-security]
[  322s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, ex.asUserString().c_str() );
[  322s]                                                                                            ^
[  323s] pk-backend-zypp.cpp: In function 'gboolean zypp_perform_execution(PkBackendJob*, zypp::ZYpp::Ptr, PerformType, gboolean, PkBitfield)':
[  323s] pk-backend-zypp.cpp:1430:77: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]     pk_backend_job_error_code (job, PK_ERROR_ENUM_DEP_RESOLUTION_FAILED, emsg);
[  323s]                                                                              ^
[  323s] pk-backend-zypp.cpp:1557:91: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, ex.asUserString().c_str() );
[  323s]                                                                                            ^
[  323s] pk-backend-zypp.cpp:1559:101: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_PACKAGE_DOWNLOAD_FAILED, ex.asUserString().c_str () );
[  323s]                                                                                                      ^
[  323s] pk-backend-zypp.cpp:1561:91: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, ex.asUserString().c_str() );
[  323s]                                                                                            ^
[  323s] pk-backend-zypp.cpp: In function 'gboolean zypp_refresh_cache(PkBackendJob*, zypp::ZYpp::Ptr, gboolean)':
[  323s] pk-backend-zypp.cpp:1620:90: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, e.asUserString().c_str() );
[  323s]                                                                                           ^
[  323s] pk-backend-zypp.cpp:1670:25: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    g_printf(repo_messages);
[  323s]                          ^
[  323s] pk-backend-zypp.cpp: In function 'void backend_install_files_thread(PkBackendJob*, GVariant*, gpointer)':
[  323s] pk-backend-zypp.cpp:2393:91: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, ex.asUserString().c_str() );
[  323s]                                                                                            ^
[  323s] pk-backend-zypp.cpp: In function 'void backend_repo_set_data_thread(PkBackendJob*, GVariant*, gpointer)':
[  323s] pk-backend-zypp.cpp:3328:88: error: format not a string literal and no format arguments [-Werror=format-security]
[  323s]    pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, ex.asString ().c_str ());
[  323s]                                                                                         ^
[  334s] cc1plus: some warnings being treated as errors
[  334s] Makefile:588: recipe for target 'libpk_backend_zypp_la-pk-backend-zypp.lo' failed
[  334s] make[3]: *** [libpk_backend_zypp_la-pk-backend-zypp.lo] Error 1
[  334s] make[3]: Leaving directory '/home/abuild/rpmbuild/BUILD/PackageKit-1.0.11/backends/zypp'
[  334s] Makefile:510: recipe for target 'all-recursive' failed
[  334s] make[2]: *** [all-recursive] Error 1
[  334s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/PackageKit-1.0.11/backends'
[  334s] Makefile:599: recipe for target 'all-recursive' failed
[  334s] make[1]: *** [all-recursive] Error 1
[  334s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/PackageKit-1.0.11'
[  334s] Makefile:531: recipe for target 'all' failed
[  334s] make: *** [all] Error 2
[  334s] error: Bad exit status from /var/tmp/rpm-tmp.Ip7w3M (%build)
```

Here's the complete [build log for PackageKit with zypp for Fedora 23](https://github.com/hughsie/PackageKit/files/117197/pk_with-zypp_log.txt), along with [the project on OBS where it is being built](https://build.opensuse.org/project/show/home:Pharaoh_Atem:Zypper_Fedora).

--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -275,7 +275,7 @@ struct InstallResolvableReportReceiver : public zypp::callback::ReceiveReport<zy
 	}
 
 	virtual Action problem (zypp::Resolvable::constPtr resolvable, Error error, const std::string &description, RpmLevel level) {
-		pk_backend_job_error_code (_job, PK_ERROR_ENUM_PACKAGE_FAILED_TO_INSTALL, description.c_str ());
+		pk_backend_job_error_code (_job, PK_ERROR_ENUM_PACKAGE_FAILED_TO_INSTALL, "%s", description.c_str ());
 		return ABORT;
 	}
 
@@ -310,7 +310,7 @@ struct RemoveResolvableReportReceiver : public zypp::callback::ReceiveReport<zyp
 	}
 
 	virtual Action problem (zypp::Resolvable::constPtr resolvable, Error error, const std::string &description) {
-                pk_backend_job_error_code (_job, PK_ERROR_ENUM_CANNOT_REMOVE_SYSTEM_PACKAGE, description.c_str ());
+                pk_backend_job_error_code (_job, PK_ERROR_ENUM_CANNOT_REMOVE_SYSTEM_PACKAGE, "%s", description.c_str ());
 		return ABORT;
 	}
 
@@ -409,7 +409,7 @@ struct MediaChangeReportReceiver : public zypp::callback::ReceiveReport<zypp::me
 {
 	virtual Action requestMedia (zypp::Url &url, unsigned mediaNr, const std::string &label, zypp::media::MediaChangeReport::Error error, const std::string &description, const std::vector<std::string> & devices, unsigned int &dev_current)
 	{
-		pk_backend_job_error_code (_job, PK_ERROR_ENUM_REPO_NOT_AVAILABLE, description.c_str ());
+		pk_backend_job_error_code (_job, PK_ERROR_ENUM_REPO_NOT_AVAILABLE, "%s", description.c_str ());
 		// We've to abort here, because there is currently no feasible way to inform the user to insert/change media
 		return ABORT;
 	}
@@ -599,10 +599,10 @@ ZyppJob::get_zypp()
 			initialized = TRUE;
 		}
 	} catch (const ZYppFactoryException &ex) {
-		pk_backend_job_error_code (priv->currentJob, PK_ERROR_ENUM_FAILED_INITIALIZATION, ex.asUserString().c_str() );
+		pk_backend_job_error_code (priv->currentJob, PK_ERROR_ENUM_FAILED_INITIALIZATION, "%s", ex.asUserString().c_str() );
 		return NULL;
 	} catch (const Exception &ex) {
-		pk_backend_job_error_code (priv->currentJob, PK_ERROR_ENUM_INTERNAL_ERROR, ex.asUserString().c_str() );
+		pk_backend_job_error_code (priv->currentJob, PK_ERROR_ENUM_INTERNAL_ERROR, "%s", ex.asUserString().c_str() );
 		return NULL;
 	}
 
@@ -952,7 +952,7 @@ zypp_get_Repository (PkBackendJob *job, const gchar *alias)
 		RepoManager manager;
 		info = manager.getRepositoryInfo (alias);
 	} catch (const repo::RepoNotFoundException &ex) {
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, ex.asUserString().c_str() );
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, "%s", ex.asUserString().c_str() );
 		return RepoInfo ();
 	}
 
@@ -1402,7 +1402,7 @@ zypp_perform_execution (PkBackendJob *job, ZYpp::Ptr zypp, PerformType type, gbo
 					it->statusReset ();
 			}
 
-			pk_backend_job_error_code (job, PK_ERROR_ENUM_DEP_RESOLUTION_FAILED, emsg);
+			pk_backend_job_error_code (job, PK_ERROR_ENUM_DEP_RESOLUTION_FAILED, "%s", emsg);
 			g_free (emsg);
 
 			goto exit;
@@ -1529,11 +1529,11 @@ zypp_perform_execution (PkBackendJob *job, ZYpp::Ptr zypp, PerformType type, gbo
 		pk_backend_job_set_percentage(job, 100);
 		ret = TRUE;
 	} catch (const repo::RepoNotFoundException &ex) {
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, ex.asUserString().c_str() );
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, "%s", ex.asUserString().c_str() );
 	} catch (const target::rpm::RpmException &ex) {
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_PACKAGE_DOWNLOAD_FAILED, ex.asUserString().c_str () );
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_PACKAGE_DOWNLOAD_FAILED, "%s", ex.asUserString().c_str () );
 	} catch (const Exception &ex) {
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, ex.asUserString().c_str() );
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, "%s", ex.asUserString().c_str() );
 	}
 
  exit:
@@ -1592,7 +1592,7 @@ zypp_refresh_cache (PkBackendJob *job, ZYpp::Ptr zypp, gboolean force)
 	catch ( const Exception &e)
 	{
 		// FIXME: make sure this dumps out the right sring.
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, e.asUserString().c_str() );
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, "%s", e.asUserString().c_str() );
 		return FALSE;
 	}
 
@@ -1642,7 +1642,7 @@ zypp_refresh_cache (PkBackendJob *job, ZYpp::Ptr zypp, gboolean force)
 		pk_backend_job_set_percentage (job, i >= num_of_repos ? 100 : (100 * i) / num_of_repos);
 	}
 	if (repo_messages != NULL)
-		g_printf(repo_messages);
+		g_printf("%s", repo_messages);
 
 	pk_backend_job_set_percentage (job, 100);
 	g_free (repo_messages);
@@ -2365,7 +2365,7 @@ backend_install_files_thread (PkBackendJob *job, GVariant *params, gpointer user
 	try {
 		manager.removeRepository (tmpRepo);
 	} catch (const repo::RepoNotFoundException &ex) {
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, ex.asUserString().c_str() );
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_REPO_NOT_FOUND, "%s", ex.asUserString().c_str() );
 	}
 }
 
@@ -3300,7 +3300,7 @@ backend_repo_set_data_thread (PkBackendJob *job, GVariant *params, gpointer user
 	} catch (const repo::RepoException &ex) {
 		pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, "Can't access the given URL");
 	} catch (const Exception &ex) {
-		pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, ex.asString ().c_str ());
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR, "%s", ex.asString ().c_str ());
 	}
 }
 

